File: plugins/motherduck/skills/motherduck-build-dashboard/artifacts/dashboard_story_example.py
import json
import sys
from pathlib import Path

import duckdb

sys.path.append(str(Path(__file__).resolve().parents[3]))

from scripts._lib.motherduck_artifact_utils import artifact_session


def one(conn: duckdb.DuckDBPyConnection, sql: str) -> dict:
    cursor = conn.execute(sql)
    columns = [col[0] for col in cursor.description]
    row = cursor.fetchone()
    return dict(zip(columns, row))


def many(conn: duckdb.DuckDBPyConnection, sql: str) -> list[dict]:
    cursor = conn.execute(sql)
    columns = [col[0] for col in cursor.description]
    return [dict(zip(columns, row)) for row in cursor.fetchall()]


def main() -> None:
    with artifact_session(slug="motherduck-build-dashboard", database_keys=["analytics"]) as session:
        conn = session.conn
        orders_table = session.table("analytics", "main", "orders")
        conn.execute(
            f"""
            CREATE TABLE {orders_table} (
                order_id INTEGER,
                order_date DATE,
                category VARCHAR,
                customer_id INTEGER,
                revenue DOUBLE
            )
            """
        )
        conn.executemany(
            f"INSERT INTO {orders_table} VALUES (?, ?, ?, ?, ?)",
            [
                (1, "2026-01-03", "Database", 101, 1200.0),
                (2, "2026-01-07", "Compute", 102, 800.0),
                (3, "2026-02-11", "Database", 101, 1600.0),
                (4, "2026-02-21", "Sharing", 103, 400.0),
                (5, "2026-03-03", "Compute", 104, 2200.0),
                (6, "2026-03-18", "Database", 105, 900.0),
            ],
        )

        result = {
            "backend": session.describe(),
            "story": "Revenue and product mix",
            "kpis": one(
                conn,
                f"""
                SELECT
                  SUM(revenue) AS total_revenue,
                  COUNT(DISTINCT order_id) AS order_count,
                  COUNT(DISTINCT customer_id) AS customer_count
                FROM {orders_table}
                """,
            ),
            "trend": many(
                conn,
                f"""
                SELECT strftime(date_trunc('month', order_date), '%Y-%m') AS month,
                       SUM(revenue) AS revenue
                FROM {orders_table}
                GROUP BY 1
                ORDER BY 1
                """,
            ),
            "breakdown": many(
                conn,
                f"""
                SELECT category, SUM(revenue) AS revenue
                FROM {orders_table}
                GROUP BY 1
                ORDER BY revenue DESC
                """,
            ),
            "detail": many(
                conn,
                f"""
                SELECT strftime(order_date, '%Y-%m-%d') AS order_date,
                       category,
                       revenue
                FROM {orders_table}
                ORDER BY order_date DESC
                LIMIT 5
                """,
            ),
        }
        print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: plugins/motherduck/skills/motherduck-build-dashboard/artifacts/dashboard_story_example.ts
export {};
declare const process: { env: Record<string, string | undefined> };

type OrderRow = {
  order_id: number;
  order_date: string;
  category: string;
  customer_id: number;
  revenue: number;
};

function normalizeMetadataValue(value: string | undefined, fallback: string): string {
  const raw = (value ?? "").trim();
  if (!raw) return fallback;
  const normalized = raw.replace(/[^A-Za-z0-9._-]+/g, "-").replace(/^[-._]+|[-._]+$/g, "");
  return normalized || fallback;
}

function buildUseCaseUserAgent(): string {
  const harness = normalizeMetadataValue(process.env.MOTHERDUCK_AGENT_HARNESS, "unknown");
  const llm = normalizeMetadataValue(process.env.MOTHERDUCK_AGENT_LLM, "unknown");
  return `agent-skills/2.2.2(harness-${harness};llm-${llm})`;
}

const orders: OrderRow[] = [
  { order_id: 1, order_date: "2026-01-03", category: "Database", customer_id: 101, revenue: 1200.0 },
  { order_id: 2, order_date: "2026-01-07", category: "Compute", customer_id: 102, revenue: 800.0 },
  { order_id: 3, order_date: "2026-02-11", category: "Database", customer_id: 101, revenue: 1600.0 },
  { order_id: 4, order_date: "2026-02-21", category: "Sharing", customer_id: 103, revenue: 400.0 },
  { order_id: 5, order_date: "2026-03-03", category: "Compute", customer_id: 104, revenue: 2200.0 },
  { order_id: 6, order_date: "2026-03-18", category: "Database", customer_id: 105, revenue: 900.0 },
];

const totalRevenue = orders.reduce((sum, row) => sum + row.revenue, 0);
const orderCount = new Set(orders.map((row) => row.order_id)).size;
const customerCount = new Set(orders.map((row) => row.customer_id)).size;

const trendMap = new Map<string, number>();
for (const row of orders) {
  const month = row.order_date.slice(0, 7);
  trendMap.set(month, (trendMap.get(month) ?? 0) + row.revenue);
}
const trend = Array.from(trendMap.entries())
  .map(([month, revenue]) => ({ month, revenue }))
  .sort((a, b) => a.month.localeCompare(b.month));

const breakdownMap = new Map<string, number>();
for (const row of orders) {
  breakdownMap.set(row.category, (breakdownMap.get(row.category) ?? 0) + row.revenue);
}
const breakdown = Array.from(breakdownMap.entries())
  .map(([category, revenue]) => ({ category, revenue }))
  .sort((a, b) => b.revenue - a.revenue);

const detail = [...orders]
  .sort((a, b) => b.order_date.localeCompare(a.order_date))
  .slice(0, 5)
  .map(({ order_date, category, revenue }) => ({ order_date, category, revenue }));

const result = {
  backend: {
    mode: "typescript-companion",
    databases: { analytics: "analytics" },
    user_agent: buildUseCaseUserAgent(),
  },
  story: "Revenue and product mix",
  kpis: {
    total_revenue: totalRevenue,
    order_count: orderCount,
    customer_count: customerCount,
  },
  trend,
  breakdown,
  detail,
};

console.log(JSON.stringify(result, null, 2));
